From 2cf63f7925eaf00f184e532ca6fd30a224a7d668 Mon Sep 17 00:00:00 2001
From: Ken Tseng <ken.tseng@realtek.com>
Date: Tue, 20 Jan 2026 15:25:51 +0800
Subject: [PATCH] Bluetooth: Support Realtek RTL8822ES UART BT

- dts: rtd1619b-backinblack: add bluetooth in uart1
- btrtl: load RTL8822ES FWs in rtlbt
- hci_h5: check CONFIG_BT_RTKBTRFKILL
-- set h5->enable_gpio to bt_reset on EBUSY
-- call rtk_voltage_detect_and_set()
---
 .../dts/realtek/rtd1619b-backinblack.dtsi     |  5 ++
 drivers/bluetooth/btrtl.c                     | 12 +++
 drivers/bluetooth/hci_h5.c                    | 20 +++++
 4 files changed, 83 insertions(+), 33 deletions(-)

diff --git a/arch/arm64/boot/dts/realtek/rtd1619b-backinblack.dtsi b/arch/arm64/boot/dts/realtek/rtd1619b-backinblack.dtsi
index 535975054502b..13545ca65602a 100644
--- a/arch/arm64/boot/dts/realtek/rtd1619b-backinblack.dtsi
+++ b/arch/arm64/boot/dts/realtek/rtd1619b-backinblack.dtsi
@@ -478,6 +478,11 @@ &uart0 {
 /* M.2 slot (CON2) */
 &uart1 {
 	status = "okay";
+
+	bluetooth {
+		compatible = "realtek,rtl8822es-bt";
+		enable-gpios = <&gpio 6 GPIO_ACTIVE_HIGH>;
+	};
 };
 
 &uart2 {
diff --git a/drivers/bluetooth/btrtl.c b/drivers/bluetooth/btrtl.c
index 812f6941b2633..3f3699a5ced48 100644
--- a/drivers/bluetooth/btrtl.c
+++ b/drivers/bluetooth/btrtl.c
@@ -246,6 +246,15 @@ static const struct id_table ic_id_table[] = {
 	  .cfg_name = "rtl_bt/rtl8822cs_config",
 	  .hw_info  = "rtl8822cs" },
 
+	/* 8822E with UART interface */
+	{ IC_INFO(RTL_ROM_LMP_8822B, 0xe, 0xc, HCI_UART),
+	  .config_needed = true,
+	  .has_rom_version = true,
+	  .has_msft_ext = false,
+	  .fw_name  = "rtlbt/rtl8822es_fw",
+	  .cfg_name = "rtlbt/rtl8822es_config",
+	  .hw_info  = "rtl8822es" },
+
 	/* 8822C with USB interface */
 	{ IC_INFO(RTL_ROM_LMP_8822B, 0xc, 0xa, HCI_USB),
 	  .config_needed = false,
@@ -654,6 +663,7 @@ static int rtlbt_parse_firmware(struct hci_dev *hdev,
 		{ RTL_ROM_LMP_8852A, 18 },	/* 8852A */
 		{ RTL_ROM_LMP_8852A, 20 },	/* 8852B */
 		{ RTL_ROM_LMP_8852A, 25 },	/* 8852C */
+		{ RTL_ROM_LMP_8822B, 33 },	/* 8822E */
 		{ RTL_ROM_LMP_8851B, 36 },	/* 8851B */
 		{ RTL_ROM_LMP_8852A, 47 },	/* 8852BT */
 	};
@@ -1549,6 +1559,8 @@ MODULE_FIRMWARE("rtl_bt/rtl8822b_fw.bin");
 MODULE_FIRMWARE("rtl_bt/rtl8822b_config.bin");
 MODULE_FIRMWARE("rtl_bt/rtl8822cs_fw.bin");
 MODULE_FIRMWARE("rtl_bt/rtl8822cs_config.bin");
+MODULE_FIRMWARE("rtlbt/rtl8822es_fw.bin");
+MODULE_FIRMWARE("rtlbt/rtl8822es_config.bin");
 MODULE_FIRMWARE("rtl_bt/rtl8822cu_fw.bin");
 MODULE_FIRMWARE("rtl_bt/rtl8822cu_config.bin");
 MODULE_FIRMWARE("rtl_bt/rtl8851bu_fw.bin");
diff --git a/drivers/bluetooth/hci_h5.c b/drivers/bluetooth/hci_h5.c
index c0436881a533c..da284bdcf43b1 100644
--- a/drivers/bluetooth/hci_h5.c
+++ b/drivers/bluetooth/hci_h5.c
@@ -104,6 +104,7 @@ struct h5 {
 
 enum h5_driver_info {
 	H5_INFO_WAKEUP_DISABLE = BIT(0),
+	H5_VOLT_DETECT_LSADC = BIT(1),
 };
 
 struct h5_vnd {
@@ -816,6 +817,10 @@ static const struct hci_uart_proto h5p = {
 	.flush		= h5_flush,
 };
 
+#if IS_ENABLED(CONFIG_BT_RTKBTRFKILL)
+extern struct gpio_desc *bt_reset;
+void rtk_voltage_detect_and_set(int method, int mode);
+#endif
 static int h5_serdev_probe(struct serdev_device *serdev)
 {
 	struct device *dev = &serdev->dev;
@@ -856,6 +861,9 @@ static int h5_serdev_probe(struct serdev_device *serdev)
 		set_bit(H5_WAKEUP_DISABLE, &h5->flags);
 
 	h5->enable_gpio = devm_gpiod_get_optional(dev, "enable", GPIOD_OUT_LOW);
+#if IS_ENABLED(CONFIG_BT_RTKBTRFKILL)
+	if (IS_ERR(h5->enable_gpio) && (PTR_ERR(h5->enable_gpio) == -EBUSY)) h5->enable_gpio = bt_reset;
+#endif
 	if (IS_ERR(h5->enable_gpio))
 		return PTR_ERR(h5->enable_gpio);
 
@@ -864,6 +872,11 @@ static int h5_serdev_probe(struct serdev_device *serdev)
 	if (IS_ERR(h5->device_wake_gpio))
 		return PTR_ERR(h5->device_wake_gpio);
 
+#if IS_ENABLED(CONFIG_BT_RTKBTRFKILL)
+	if (data->driver_info & H5_VOLT_DETECT_LSADC)
+		rtk_voltage_detect_and_set(1, 0);
+#endif
+
 	return hci_uart_register_device_priv(&h5->serdev_hu, &h5p,
 					     h5->vnd->sizeof_priv);
 }
@@ -1079,6 +1092,11 @@ static const struct h5_device_data h5_data_rtl8822cs = {
 	.vnd = &rtl_vnd,
 };
 
+static const struct h5_device_data h5_data_rtl8822es = {
+	.driver_info = H5_VOLT_DETECT_LSADC,
+	.vnd = &rtl_vnd,
+};
+
 static const struct h5_device_data h5_data_rtl8723bs = {
 	.driver_info = H5_INFO_WAKEUP_DISABLE,
 	.vnd = &rtl_vnd,
@@ -1105,6 +1123,8 @@ static const struct of_device_id rtl_bluetooth_of_match[] = {
 #ifdef CONFIG_BT_HCIUART_RTL
 	{ .compatible = "realtek,rtl8822cs-bt",
 	  .data = (const void *)&h5_data_rtl8822cs },
+	{ .compatible = "realtek,rtl8822es-bt",
+	  .data = (const void *)&h5_data_rtl8822es },
 	{ .compatible = "realtek,rtl8723bs-bt",
 	  .data = (const void *)&h5_data_rtl8723bs },
 	{ .compatible = "realtek,rtl8723cs-bt",
-- 
2.34.1

